File: quicksort/python/quicksort.py
#!/usr/bin/python


def quicksort(lst):
    l = len(lst)
    if l <= 1:
        return lst
    pivot = lst[l // 2]
    equal = []
    lesser = []
    greater = []
    for e in lst:
        if e < pivot:
            lesser.append(e)
        elif e > pivot:
            greater.append(e)
        else:
            equal.append(e)
    if greater:
        greater = quicksort(greater)
    if lesser:
        lesser = quicksort(lesser)
    return lesser + equal + greater


lst = list(range(10, 0, -1))
print(lst)
print(quicksort(lst))

File: quicksort/c++/quicksort.cpp
#include <iostream>
#include <vector>

using std::vector;
using std::cout;
using std::endl;

template<typename T>
T quicksort( T arr ) {
	if( arr.size() <= 1 ) {
		return arr;
	}

	T equal;
	T lesser;
	T greater;

	auto pivot = arr[ arr.size() / 2 ];

	for( auto && e : arr ) {
		if( e < pivot ) {
			lesser.push_back( e );
		} else if( e > pivot ) {
			greater.push_back( e );
		} else {
			equal.push_back( e );
		}
	}

	if( !greater.empty() ) {
		greater = quicksort( greater );
	}

	if( !lesser.empty() ) {
		lesser = quicksort( lesser );
	}

	T result;
	result.reserve( arr.size() );
	result.insert( result.end(), lesser.begin(), lesser.end() );
	result.insert( result.end(), equal.begin(), equal.end() );
	result.insert( result.end(), greater.begin(), greater.end() );

	return result;
}

int main() {
	vector<int> arr = { 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 11, 3, 0 };
	auto result = quicksort( arr );
	for( auto && e : result ) {
		cout << e << ", ";
	}
	cout << endl;
}
